subprojects/core/src/integTest/groovy/org/gradle/caching/internal/packaging/BuildCacheEntryPackingIntegrationTest.groovy

```diff
@@ -0,0 +1,115 @@
+import spock.lang.Unroll
+
+class BuildCacheEntryPackingIntegrationTest extends DaemonIntegrationSpec implements DirectoryBuildCacheFixture {
+
+    private static final NON_ASCII_NAME = [
+        "ascii-only": "ascii",
+        "space": " ",
+        "zwnj": "\u200c",
+        "chinese": "敏捷的棕色狐狸跳过了懒狗",
+        "cyrillic": "здравствуйте",
+        "hungarian": "Árvíztűrő tükörfúrógép",
+    ].values().join("-")
+
+    private static final DEFAULT_ENCODINGS = [
+        "UTF-8",
+        "ISO-8859-1",
+        "windows-1250",
+    ]
+
+    @Unroll
+    def "can store and load files having non-ascii characters in file name using default file encoding #fileEncoding"() {
+        def fileName = NON_ASCII_NAME + ".txt"
+        def outputFile = file("dir", fileName)
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing file in cache: $fileName"
+
+            task createFile {
+                outputs.dir("dir")
+                outputs.cacheIf { true }
+                doLast {
+                    file("dir/$fileName").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        outputFile.text == "output"
+
+        where:
+        fileEncoding << DEFAULT_ENCODINGS
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9877")
+    @Unroll
+    def "can store and load files having non-ascii characters in property name using default file encoding #fileEncoding"() {
+        def outputFile = file("output.txt")
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing with property name: $NON_ASCII_NAME"
+
+            task createFile {
+                outputs.file("output.txt")
+                    .withPropertyName("$NON_ASCII_NAME")
+                outputs.cacheIf { true }
+                doLast {
+                    file("output.txt").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        outputFile.text == "output"
```

Comment:
❓ what's the purpose of checking output when the task was skipped?

Reply:
That's the whole point here. The task is skipped because it's loaded from cache. We need to check if the output is created as expected.

Reply:
wasn't yet aware that skipped meant it could be restored from cache. Thanks for clarifying.